prov/verbs/src/verbs_cq.c

```diff
@@ -592,12 +593,29 @@ int fi_ibv_cq_open(struct fid_domain *domain_fid, struct fi_cq_attr *attr,
 	}
 
 	size = attr->size ? attr->size : VERBS_DEF_CQ_SIZE;
+	ret = ibv_query_device(domain->verbs, &device_attr);
+	if (OFI_UNLIKELY(ret)) {
+		VERBS_INFO_ERRNO(FI_LOG_FABRIC,
+				 "ibv_query_device", errno);
+	} else {
+		if (size > device_attr.max_cqe) {
+			VERBS_WARN(FI_LOG_CQ,
+				   "Ignoring CQ size requested by user - "
+				   "%"PRIu64", because maximum supported CQ "
+				   "size by device - %d\n",
+				   size, device_attr.max_cqe);
+			size = device_attr.max_cqe;
+		}
+	}
+	VERBS_INFO(FI_LOG_CQ, "Used CQ size - %"PRIu64"\n", size);
```

Comment:
I think defining `domain_attr::cq_size` may be better option than silently changing the CQ size on `fi_cq_open`. RxM calculates the CQ size required based on TX, RX and universe size. If verbs ignores the value, that would result in dropped completions.

Reply:
RxM will eventually need to deal with these limits at scale.  It can either allocated multiple CQs as needed, or restrict the total number of posted operations in order to avoid overrunning the queues.

If we need to expose the cq_size through the domain attribute, that's reasonable.  But silently changing the size to something smaller is asking for trouble.

Reply:
Also I proposed (commit message) the second solution for this problem:
"If an user passes fi_cq_attr::size that exceeds maximum supported by HW/SW
value, a provider returns -FI_EINVAL and updates fi_cq_attr::size to the
maximum supported value. Then user may pass fi_cq_attr again, and CQ
can be allocated."

What do you think?

Reply:
Other than finding the right return value, this seems workable.

Reply:
I'm in favor of this idea, because we don't need to add mew filed in `fi_domain_attr` structure.
But I think we need update docs, that should say that user structure `fi_cq_attr` will be update in this case

Reply:
I don't see a great existing return value for this.  So, for this to work, we would need to define a new value, and only support this option for version 1.7+.  I'm not sure if that's much better than exposing hardware limits directly through some mechanism.

Reply:
Is -FI_EINVAL not good enough? If we need to support different options based on versions, we might as well add a new domain attribute. What do you think?

Reply:
I wouldn't use EINVAL for this.  We could match to a call like fi_getname() -- ETOOSMALL -- only using the opposite value.  IMO, having the limit up front would be easier for apps to adjust to, rather than relying on create/fail/try-again semantics.

I'm not sure if we extend the domain attributes or define some sort of new domain limits.  The latter would allow us to expose other hardware based limitations that are currently masked.